### DISCUSSION TITLE
Naming conventions

### DISCUSSION THREAD

**Discussion by sergiogragera** [nmanero/ddd-crud#18] [General]
The system has routes to Create, Read, Update, or Delete (CRUD) and these controllers can invoke the following operations:

1. CREATE operations could be adding an item, or inserting an item at a specific position, and so on.
2. READ operations could be get a collection or get a specific item from a collection, and so on.
3. UPDATE operations could be to replace an item or update part of its attributes, and so on.
4. DELETE operations could be to clear a collection or delete an item, and so on.

That is, we have a controller for CREATE (route name) operation that will call methods like ADD or INSERT (service layer names) and finally PERSIST (persistence layer names) this information. SAVE can also be a good name in the persistence layer, what do you think?

We will also have a controller for READ operations that will call methods like GET ALL items from collection or GET (do you prefer FIND or FETCH?) a single item from the collection and that will invoke calls to the persistence layer such as FIND ALL or FIND BY ID.

On the other hand we have UPDATE controller that will call methods like EDIT or REPLACE and invoke MERGE call from the persistence layer. UPDATE can also be a good name in the persistence layer, what do you think?

Finally we have DELETE controller that will call methods like CLEAR a collection (do you prefere DELETE ALL?) or DELETE a single item and finally invokes REMOVE operation in the persistence layer, or REMOVE ALL that will invoke a single operation in the case of SQL to optimize. 

What is your opinion about this? The nomenclature adopted in the persistence layer is based on JPA. I'm trying to solve my doubts (and be able to answer @nmanero [proposal](https://github.com/nmanero/ddd-crud/pull/14#pullrequestreview-615321452)) about naming the service layer methods.

**Comment by nmanero** [comment]
For reading, I'd prefer GETALL or GET rather than findAll or find, because as we are returning the objects, I think "get" is more accurate.
With FIND I think we would need to use both FIND and FETCH: I mean, first you find the records and then you fetch them. With GET I think that just one word means the same than those two.

About Create, I would use add or insert but not persist. Persist I think it's too close to database storage, but if we want to save it to a file, I wouldn't use persist. I think Save or Add is more generic.

Same with Update, I think Update or Edit sounds good and maybe also Replace.

About Delete or Remove, I think that both words mean "delete". If instead of deleting we are going to update a "isDeleted" column in the database, maybe we could use something like "invalidate" or "hide"

What do you think @sergiogragera @elias480 ?

**Reply by sergiogragera** [reply]
I totally agree.

On the other hand, if it were a voting system I would have one controller (route) to vote for a president and another to vote for a senator, for example. Both controllers would share a voting service with a single voting method and I would set the voting type (president, senator) as an argument and finally invoke the save method to vote repository. Therefore the system would have a controller layer with methods like "voteForPresident" or "voteForSenator", a service with "vote" method and a vote repository with "save" method.

However this system is a generic CRUD system and therefore there is no difference between adding a user (register) or adding a vote (vote) or adding a product (buy) in the name of the methods (add).

**Reply by nmanero** [reply]
Agree. If it turns into a voting system, we'll refactor the routes

**Reply by elias480** [reply]
I agree with the naming convention that you purposed @nmanero.

Personally I prefer fetch and fetch all to get or find. Many public apis use this naming and I think it's very common.

What do you prefer?

**Reply by sergiogragera** [reply]
First, nomenclatures like **persist**, **merge**, and others inherited from *JPA* have been removed from the persistence layer.
On the other hand, **get** and **getAll** have been adopted instead of **find** or **fetch** because what we are doing in those methods is **find** and **fetch** so **get** seems the best option.

I am sure there will be new proposals in the future but this is a good starting point.

**Reply by nmanero** [reply]
Fetch and get I think are similar @elias480 , but let's start with get
